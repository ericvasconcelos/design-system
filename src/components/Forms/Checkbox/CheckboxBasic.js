import React from 'react'

import { Text } from 'components/Text'
import { bool, func, string } from 'prop-types'

import { Label, StyledInput } from './styles'

export const CheckboxBasic = ({
  id,
  name,
  label,
  value,
  checked,
  indeterminate,
  onChange,
  onBlur,
  disabled,
  messageError,
  className
}) => (
  <>
    <Label htmlFor={id} className={className}>
      <StyledInput
        id={id}
        name={name}
        disabled={disabled}
        ref={el => {
          el && indeterminate && el.setAttribute('indeterminate', '')
        }}
        type="checkbox"
        {...(value !== '' ? { value } : undefined)}
        onChange={onChange}
        onBlur={onBlur}
        defaultChecked={checked}
      />
      <Text tag="span" styleType="body2" styleColor="darkBlueCash500">
        {label}
      </Text>
    </Label>
    {messageError && (
      <Text tag="p" styleType="body3" styleColor="error400">
        {messageError}
      </Text>
    )}
  </>
)

CheckboxBasic.propTypes = {
  id: string.isRequired,
  name: string.isRequired,
  label: string.isRequired,
  value: string,
  onChange: func,
  onBlur: func,
  disabled: bool,
  checked: bool,
  indeterminate: bool,
  messageError: string,
  className: string,
  labelStyleColor: string
}

CheckboxBasic.defaultProps = {
  value: '',
  disabled: false,
  labelStyleColor: 'darkBlueCash200',
  indeterminate: false,
  onChange: () => {},
  onBlur: () => {},
  checked: false,
  messageError: '',
  className: ''
}
