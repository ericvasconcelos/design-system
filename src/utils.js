export { arrayCombinations } from 'utils/arrayCombinations'
export { getBackgroundColor } from 'utils/getBackgroundColor'
export { getBorder } from 'utils/getBorder'
export { getButtonStyle } from 'utils/getButtonStyle'
export { getBorderRadius } from 'utils/getBorderRadius'
export { getColorFromName } from 'utils/getColorFromName'
export { getGuttersNames } from 'utils/getGuttersNames'
export { getIconNames } from 'utils/getIconNames'
export { getPadding } from 'utils/getPadding'
export { getPaddingNames } from 'utils/getPaddingNames'
export { getTypography } from 'utils/getTypography'
export { parsePalleteName } from 'utils/parsePalleteName'
export { styleColorValidator } from 'utils/styleColorValidator'
export { useAriaProps } from 'utils/useAriaProps'

export { DateUtils } from 'utils/DateUtils'
export { QueryUtils } from 'utils/QueryUtils'
export { ObjectUtils }  from 'utils/ObjectUtils'
export { CookiesUtils } from 'utils/CookiesUtils'
export { MoneyUtils } from 'utils/MoneyUtils'
export { NumberUtils } from 'utils/NumberUtils'