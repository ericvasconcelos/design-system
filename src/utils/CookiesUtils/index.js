export class CookiesUtils {
  static toObject(string = '') {
    const cookies = {}
    if (string !== '') {
      const parts = string.split('; ')
      parts.forEach(part => {
        const [key, value] = part.split('=')
        cookies[key] = value
      })
    }
    return cookies
  }
}
