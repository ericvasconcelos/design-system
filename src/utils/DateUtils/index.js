export class DateUtils {
  static toDate(string, format = 'pt-br') {
    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = string?.split('/')
        break
      case 'iso':
        [year, month, day] = string?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    return new Date(year, parseInt(month, 10) - 1, day)
  }

  static getAge(date) {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const difference = Date.now() - date.getTime()

    return Math.abs(new Date(difference).getUTCFullYear() - 1970)
  }

  static toString(date, format = 'pt-br') {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
    const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1
    const year = date.getFullYear()

    switch (format) {
      case 'pt-br':
        return `${day}/${month}/${year}`
      case 'iso':
        return `${year}-${month}-${day}`
      default:
        throw new Error('Format invalid')
    }
  }

  static validate(date, format = 'pt-br') {
    const monthDaysMap = {
      '01': 31,
      '02': 29,
      '03': 31,
      '04': 30,
      '05': 31,
      '06': 30,
      '07': 31,
      '08': 31,
      '09': 30,
      10: 31,
      11: 30,
      12: 31
    }
    let isValid = false

    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = date?.split('/')
        break
      case 'iso':
        [year, month, day] = date?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    isValid = isValid || parseInt(day, 10) <= monthDaysMap[month]
    isValid = isValid && parseInt(year, 10) > 1900

    return isValid
  }

  static convertMonthToYear(month) {
    return parseInt(month, 10) / 12
  }

  static formatPeriod(period, periodType) {
    const periodTypeMap = {
      monthly: { singular: 'mÃªs', plural: 'meses' },
      yearly: { singular: 'ano', plural: 'anos' },
      daily: { singular: 'dia', plural: 'dias' }
    }
    return `${period} ${periodTypeMap[periodType][period < 2 ? 'singular' : 'plural']}`
  }
}
