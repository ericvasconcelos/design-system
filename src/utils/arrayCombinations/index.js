export const arrayCombinations = (...args) => {
  const returnable = []
  const max = args.length - 1
  function helper(arr, i) {
    for (let j = 0, { length } = args[i]; j < length; j++) {
      const newArray = arr.slice(0) // clone arr
      newArray.push(args[i][j])
      if (i === max) returnable.push(newArray)
      else helper(newArray, i + 1)
    }
  }
  helper([], 0)
  return returnable
}
