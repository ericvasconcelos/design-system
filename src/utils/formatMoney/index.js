import { toMoney, toNumber } from 'vanilla-masker'

export const formatMoney = entry => {
  const moneyFormatter = {
    precision: 2,
    separator: ',',
    delimiter: '.',
    unit: 'R$'
  }

  return toMoney(toNumber(entry), moneyFormatter)
}
