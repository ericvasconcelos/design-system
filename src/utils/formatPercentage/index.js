import { toMoney, toNumber } from 'vanilla-masker'

export const formatPercentage = entry => {
  const percentageFormatter = {
    precision: 1,
    separator: ',',
    suffixUnit: '%'
  }
  return toMoney(toNumber(entry), percentageFormatter)
}
