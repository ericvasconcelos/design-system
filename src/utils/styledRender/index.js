import React from 'react'

import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme } from 'styles/theme'

export const styledRender = (children, options = {}) => {
  const rendered = render(<ThemeProvider theme={theme}>{children}</ThemeProvider>, options)
  return {
    ...rendered,
    rerender: (ui, rerenderOptions = {}) =>
      styledRender(ui, { container: rendered.container, ...rerenderOptions })
  }
}
