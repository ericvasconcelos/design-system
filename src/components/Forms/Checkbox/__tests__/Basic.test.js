import React from 'react'

import userEvent from '@testing-library/user-event'
import { styledRender } from 'utils/styledRender'

import { CheckboxBasic } from '../CheckboxBasic'

describe('should render the component correctly when', () => {
  it('has basic props', () => {
    const { container, getByText, getByLabelText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" />
    )

    expect(getByLabelText('Checkbox')).not.toBeChecked()
    expect(getByText('Checkbox')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has checked prop', () => {
    const { container, getByLabelText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" checked />
    )

    expect(getByLabelText('Checkbox')).toBeChecked()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has messageError prop', () => {
    const { container, getByText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" messageError="Erro" />
    )

    expect(getByText('Erro')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has disabled prop', () => {
    const onChange = jest.fn()
    const { container, getByLabelText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" disabled onChange={onChange} />
    )

    userEvent.click(getByLabelText('Checkbox'))

    expect(onChange).not.toBeCalled()
    expect(getByLabelText('Checkbox')).toHaveAttribute('disabled')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has indeterminate prop', () => {
    const { container, getByLabelText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" indeterminate />
    )

    expect(getByLabelText('Checkbox')).toHaveAttribute('indeterminate')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('click call onChange prop', () => {
    const onChange = jest.fn()
    const { getByLabelText } = styledRender(
      <CheckboxBasic id="checkbox" name="checkbox" label="Checkbox" onChange={onChange} />
    )

    userEvent.click(getByLabelText('Checkbox'))

    expect(getByLabelText('Checkbox')).toBeChecked()
    expect(onChange).toBeCalled()
  })
})
