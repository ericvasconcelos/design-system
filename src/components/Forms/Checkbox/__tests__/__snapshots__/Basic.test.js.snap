// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component correctly when has basic props 1`] = `
.c2 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c1[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c1[checked] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #FFFFFF;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<label
  class="c0"
  for="checkbox"
>
  <input
    class="c1"
    id="checkbox"
    name="checkbox"
    type="checkbox"
  />
  <span
    class="c2"
    data-testid="Text"
  >
    Checkbox
  </span>
</label>
`;

exports[`should render the component correctly when has checked prop 1`] = `
.c2 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c1[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c1[checked] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #FFFFFF;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<label
  class="c0"
  for="checkbox"
>
  <input
    checked=""
    class="c1"
    id="checkbox"
    name="checkbox"
    type="checkbox"
  />
  <span
    class="c2"
    data-testid="Text"
  >
    Checkbox
  </span>
</label>
`;

exports[`should render the component correctly when has disabled prop 1`] = `
.c2 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #69788C14;
  border: 2px solid #69788C14;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c1[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c1[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #69788C52;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<label
  class="c0"
  for="checkbox"
>
  <input
    class="c1"
    disabled=""
    id="checkbox"
    name="checkbox"
    type="checkbox"
  />
  <span
    class="c2"
    data-testid="Text"
  >
    Checkbox
  </span>
</label>
`;

exports[`should render the component correctly when has indeterminate prop 1`] = `
.c2 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c1[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c1[checked] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #FFFFFF;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<label
  class="c0"
  for="checkbox"
>
  <input
    class="c1"
    id="checkbox"
    indeterminate=""
    name="checkbox"
    type="checkbox"
  />
  <span
    class="c2"
    data-testid="Text"
  >
    Checkbox
  </span>
</label>
`;

exports[`should render the component correctly when has messageError prop 1`] = `
.c2 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c1[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c1[checked] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c1[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #FFFFFF;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<label
  class="c0"
  for="checkbox"
>
  <input
    class="c1"
    id="checkbox"
    name="checkbox"
    type="checkbox"
  />
  <span
    class="c2"
    data-testid="Text"
  >
    Checkbox
  </span>
</label>
`;
